### PR TITLE
Rewrite scanner service

### PR DESCRIPTION
**Pull request by mrmlnc**
#### What is the purpose of this pull request?

Update the scanner service.

#### What changes did you make? (Give an overview)

* Rectification of recursion into a loop to improve readability and maintainability. Previously, there was complex and nested recursion here.
* Correct processing of partial files. Previously, we do not check the existence of a partial file.
* Apparently, we're scanning better now.
* Previously, we scan all files after each rename — now only those that have changed.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -53,11 +53,6 @@
 					"default": true,
 					"description": "Allows scan imported files."
 				},
-				"scss.scanImportedFilesDepth": {
```
Comment (octref) [inline package.json]: I like removal of this option. It doesn't make much sense to me.

You probably want to update README as well.

**Comment by mrmlnc** [comment]
Of course, I will return this functionality.

File: src/server.ts [hunk 2]
```diff
@@ -83,24 +90,14 @@ connection.onDidChangeConfiguration(params => {
...
-		return;
-	}
+	const files = event.changes
+		.filter(file => file.type === FileChangeType.Changed || file.type === FileChangeType.Created)
+		.map(file => Files.uriToFilePath(file.uri));
 
-	return doScanner(workspaceRoot, cache, settings).then(symbols => {
-		return invalidateCacheStorage(cache, symbols);
-	});
+	return scannerService.scan(files);
```
Comment (mrmlnc) [inline src/server.ts]: We have to disable recursive scanning here because we only check changed files.

File: src/services/scanner.ts [hunk 3]
```diff
@@ -1,206 +1,70 @@
...
+export default class ScannerService {
+	constructor(private readonly _cache: ICache, private readonly _settings: ISettings) {}
 
-interface IFile {
-	filepath: string;
-	dir: string;
-	ctime: Date;
-}
+	public async scan(files: string[], recursive = true): Promise<void> {
+		const iterator = new Set([...files]);
```
Comment (mrmlnc) [inline src/services/scanner.ts]: There may be problems (Maximum call stack) if the number of files is large.

File: src/server.ts [hunk 4]
```diff
@@ -80,25 +89,15 @@ connection.onDidChangeConfiguration(params => {
...
-	}
+connection.onDidChangeWatchedFiles(event => {
+	const files = event.changes
+		.filter(file => file.type === FileChangeType.Changed || file.type === FileChangeType.Created)
+		.map(file => Files.uriToFilePath(file.uri));
 
-	const symbols = await doScanner(workspaceRoot, cache, settings);
-
-	invalidateCacheStorage(cache, symbols);
+	return scannerService.scan(files, /* recursive */ false);
```
Comment (mrmlnc) [inline src/server.ts]: We must return logic about the current document.

File: src/test/e2e/suite/completion/completion.test.ts [hunk 5]
```diff
@@ -8,7 +8,7 @@ describe('SCSS Completion Test', () => {
 
 	before(async () => {
 		showFile(docUri);
-		await sleep(5000);
+		await sleep(10000);
```
Comment (mrmlnc) [inline src/test/e2e/suite/completion/completion.test.ts]: Locally, the test runs with increasing timeout. I will  invest in it.

**Comment by mrmlnc** [comment]
@octref, please, take a look at this pull request.

File: src/services/scanner.ts [hunk 6]
```diff
@@ -1,206 +1,72 @@
...
-	}
+			if (!isExistFile && !isPartialFile) {
+				filepath = partialFilepath;
+				isExistFile = await this._fileExists(filepath);
+			}
 
-	return true;
-}
+			if (!isExistFile) {
+				this._cache.drop(filepath);
```
Comment (mrmlnc) [inline src/services/scanner.ts]: `filepath` → `originalFilepath`

File: src/server.ts (hunk 4 above)
Comment (mrmlnc) [inline src/server.ts]: Most likely here we need to always use `recursive` mode, because a new import may appear in the document.

**Comment by octref** [comment]
I had a rough look. The change looks good!
I suggest publishing a new version soon and see if user reports any bug.

**Comment by mrmlnc** [comment]
Yeap, I want to publish the new version on this weekend. With bug fix for https://github.com/mrmlnc/vscode-scss/issues/92#issuecomment-557662731

File: src/providers/completion.ts [hunk 7]
```diff
@@ -163,8 +163,8 @@ export function doCompletion(
 	// Variables
 	if (settings.suggestVariables && isVariableContext) {
 		symbolsList.forEach(symbols => {
-			const fsPath = getDocumentPath(documentPath, symbols.document);
 			const isImplicitlyImport = isImplicitly(symbols.document, documentPath, documentImports);
+			const fsPath = getDocumentPath(documentPath, isImplicitlyImport ? symbols.filepath : symbols.document);
```
Comment (mrmlnc) [inline src/providers/completion.ts]: Annotation 2019-11-23 001728

Annotation 2019-11-23 001728qwe

**Comment by octref** [comment]
One thing we could do is to remove duplicates at the end, right before returning completion, what do you think?

**Comment by mrmlnc** [comment]
Yes, we can do it.

The duplicates in the screenshot above looks like a bug. I will investigate this soon.

**Comment by mrmlnc** [comment]
UPD. This is a bug inside the `scss-symbols-parser` package.

```
[                                                                   
  { name: '$caret-width', value: '123', offset: 22 },                // variable
  { name: '$caret-width', value: 'solid transparent', offset: 91 },  // variable inside mixin
  { name: '$caret-width', value: 'solid', offset: 139 },             // variable inside mixin
  { name: '$caret-width', value: 'solid transparent', offset: 173 }  // variable inside mixin
]                                                                   
```

**Pull request merged**